### PR TITLE
PR on New Feature into Develop

### PR DESCRIPTION
**Pull request by tadeuasarro**
##### Added features:
- Sign up and log in users
- Create events
- Invite users to events
- Check events that have happened and events that are going to happen
- Check which users are invited to those events
- Check the events that are being hosted in a users show page
- Create relations between User and Event models (hosting_user)
- Create relations between User and Event models through the Invitation model (invited_user)

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hello @tadeuasarro, @Shaqri... Good start on this project. Project works well with a neat interface...:+1: But some implementation changes are required::**

:point_right: In your README, add instructions on how to setup the app and how to use it.

:point_right: Your events create action should make use of the build method. i.e. `current_user.events.build(event_params)`



:point_right: Your events index action should take care of retrieving the events into different collections for past and upcoming, and then pass the collections to the views. So you can have something like `@past_events = Event.previous_events` and `@upcoming_events = Event.upcoming_events` in the controller and then used directly in the view.


:point_right: In your user show action, use the event upcoming and past scopes to group the events instead of having an iteration manually through the collection of events. Also, have a collection for the user's created events in the show action and now retrieve this collection directly in the view. Say... `@created_events =  current_user.events`


:point_right: Add indexes for all foreign keys. See why this is necessary [here](https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html)
and [here](https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys)

:point_right: Tests are missing. Write unit tests for model validations and associations (RSpec) and integration tests for authentication and event management (RSpec + Capybara)



:point_right: Remove commented code in Events controller

_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested:exclamation:

**Review by lucasmazo32 — CHANGES_REQUESTED** [review]
# Changes required :wrench:

You are doing a great job so far, and the project is looking good. 

For the project to get approved kindly implement these changes:

## Required changes:

- [ ] I can not run the tests because of a migration error. Kindly check the log below, fix the error, and submit for another code review.

### Happy coding! :computer:

## Lucas Mazo

```
StandardError: An error has occurred, this and all later migrations canceled:

SQLite3::SQLException: table "events" already exists: CREATE TABLE "events" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar, "location" varchar, "date" date, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL)
$HOME/private-event/db/migrate/20200929174819_create_events.rb:3:in `change'
$HOME/private-event/bin/rails:9:in `<top (required)>'
$HOME/private-event/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'

Caused by:
ActiveRecord::StatementInvalid: SQLite3::SQLException: table "events" already exists: CREATE TABLE "events" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar, "location" varchar, "date" date, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL)
$HOME/private-event/db/migrate/20200929174819_create_events.rb:3:in `change'
$HOME/private-event/bin/rails:9:in `<top (required)>'
$HOME/private-event/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'

Caused by:
SQLite3::SQLException: table "events" already exists
$HOME/private-event/db/migrate/20200929174819_create_events.rb:3:in `change'
$HOME/private-event/bin/rails:9:in `<top (required)>'
$HOME/private-event/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

File: app/views/events/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    <% @upcoming_events.each do |event| %>
+      <div class="card pb-0 mb-3">
+        <div class="card-body pb-0 text-dark">
+          <h5 class="card-title">Event: <%= event.name %></h5>
+          <p class="card-text">Location: <%= event.location %><p>
+          <p class="card-text">Date: <%= event.date %><p>
+          <%= link_to event.hosting_user.username.capitalize, user_path(event.hosting_user), class: 'badge badge-light' %>
+          <%= link_to 'Details', event_path(event), class: 'badge badge-dark' %>
+        </div>
+      </div>
```
Comment (BrittanyBlake) [inline app/views/events/index.html.erb]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/events/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    <% end %>
```
Comment (BrittanyBlake) [inline app/views/events/index.html.erb]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/events/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  <div class="w-50">
+    <h3>List of invited people:</h3>
+    <% @event.invited_users.each do |iu| %>
+      <div class="card mb-3">
+        <div class="card-body">
+          <h5 class="card-title"><%= iu.username.capitalize %></h5>
+          <h6 class="card-subtitle"><%= iu.email %></h6>
+        </div>
+      </div>
+    <% end %>
```
Comment (BrittanyBlake) [inline app/views/events/show.html.erb:11]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    <% if flash[:notice]%>
+      <div class="d-flex justify-content-center p-2 alert alert-warning">
+        <p> <%= notice %> </p>
+      </div>
+    <% end %>
+    <% if flash[:alert]%>
+      <div class="d-flex justify-content-center p-2 alert alert-warning">
+        <p> <%= alert %> </p>
+      </div>
+    <% end %>
```
Comment (BrittanyBlake) [inline app/views/layouts/application.html.erb]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        <div class="mx-3">
+          <%= link_to "Home", events_path, class: 'btn btn-primary' %>
+        </div>
+        <div class="mx-3">
+          <%= link_to "Log in", sessions_new_path, class: 'btn btn-success' %>
+        </div>
+        <div class=" ml-3 mr-5">
+          <%= link_to "Sign up", new_user_path, class: 'btn btn-info' %>
+        </div>
+      <% end %>
```
Comment (BrittanyBlake) [inline app/views/layouts/application.html.erb:45]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<div class="container mt-5 d-flex justify-content-center">
+  <div class="w-25">
+    <% unless user_sign_in? %>
```
Comment (BrittanyBlake) [inline app/views/users/new.html.erb:3]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    <% end %>
+    <% @past_events.each do |event| %>
+      <div class="card pb-0 mb-3">
+        <div class="card-body pb-0 text-dark">
+          <h5 class="card-title">Event: <%= event.name %></h5>
+          <p class="card-text">Location: <%= event.location %><p>
+          <p class="card-text">Date: <%= event.date %><p>
+        </div>
+      </div>
+    <% end %>
```
Comment (BrittanyBlake) [inline app/views/users/show.html.erb]: - [ ] There should be no logic in your views. View related logic can be encapsulated in view helpers **(app/helpers/*_helper.rb)**

Screenshot from the requirememnts 👇 
Screen Shot 2020-10-05 at 11 54 11 PM

You can take a look at these links for more info on where to put your logic:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi Team! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with your project so far! It looks really good! 👏 👏 👏  There are just a few issues that need to be resolved before you can move on: ⬇️ 
****
* There should be no logic in your views. All of your view related logic should be encapsulated in helpers. I am sorry that this wasn't pointed out to you before but it is important that you get in the habit of using your helpers for view-related logic. Once you fix this your project should be ready to go! 🎖️ I have pointed out some view-related logic in the comments below! 

****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

File: app/views/events/index.html.erb (hunk 2 above)
Comment (tadeuasarro) [reply app/views/events/index.html.erb:33]: This is not an actual requirement from the project

**Review by MuriloRoque — APPROVED** [review]
# Status: Approved :green_circle: 

### Hello @tadeuasarro and @Shaqri :wave: 

### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: 

### You may merge, good luck with your next projects :rocket: 

Murilo Roque Paiva da Silva

- Github: [@muriloroque](https://github.com/MuriloRoque)
- Linkedin: [MuriloRoque](https://www.linkedin.com/in/murilo-roque-b1268741/)

[Twitter Follow](https://twitter.com/MuriloRoquePai1)

**Pull request merged**